src/compiler/checker.ts

```diff
@@ -4499,7 +4525,19 @@ namespace ts {
                         if (source === numberType && target.flags & TypeFlags.Enum) return Ternary.True;
                     }
                 }
+
+                if (relation === assignableRelation && source.flags & TypeFlags.ObjectLiteral && source.flags & TypeFlags.FreshObjectLiteral) {
+                    if (hasExcessProperties(<ObjectType>source, target, reportErrors)) {
+                        if (reportErrors) {
+                            reportRelationError(headMessage, source, target);
+                        }
+                        return Ternary.False;
+                    }
+                    source = getRegularTypeOfObjectLiteral(source);

```

Comment:
Because we otherwise will fail in cases like this:

```typescript
interface A { a: number }
interface B { b: number }
var x: A & B = { a: 1, b: 2 };
```

We make the check upfront for the entire target type, but then as we descend into the structure of the target type we no longer want to make the check again (as it would now fail).